### PR TITLE
Persist WorkflowToken to MDS

### PR DESCRIPTION
**Pull request by bdmogal**
Added ability to persist WorkflowToken to MDS after every node's execution completes.

The WorkflowToken is stored as a json in a Workflow's RunRecord's properties with the key 'workflowToken'.
Also adds two HTTP endpoints for querying workflow tokens.

Issues addressed:
[CDAP-2713](https://issues.cask.co/browse/CDAP-2713) 
[CDAP-2714](https://issues.cask.co/browse/CDAP-2714)

Build:
https://builds.cask.co/browse/CDAP-DUT2206/

Includes #3128. The changes in this PR are in the last commit.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java [hunk 1]
```diff
@@ -284,6 +268,36 @@ public void run() {
       executor.shutdownNow();
       status.remove(node.getNodeId());
     }
+    Id.Workflow workflowId = Id.Workflow.from(program.getId().getApplication(), workflowSpec.getName());
+    store.updateWorkflowToken(workflowId, runId.getId(), token);
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java]: Token values added in the condition node will not get updated right?

File: cdap-proto/src/main/java/co/cask/cdap/proto/codec/WorkflowTokenNodeDetailCodec.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonSerializationContext;
+
+import java.lang.reflect.Type;
+import java.util.Map;
+
+/**
+ * Codec to serialize/deserialize {@link WorkflowTokenNodeDetail}
```
Comment (sagarkapare) [inline cdap-proto/src/main/java/co/cask/cdap/proto/codec/WorkflowTokenNodeDetailCodec.java]: `.` after the comment.

File: cdap-proto/src/main/java/co/cask/cdap/proto/codec/WorkflowTokenDetailCodec.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonSerializationContext;
+
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Codec to serialize/deserialize {@link WorkflowTokenDetail}
```
Comment (sagarkapare) [inline cdap-proto/src/main/java/co/cask/cdap/proto/codec/WorkflowTokenDetailCodec.java]: `.` after the comment.

File: cdap-proto/src/main/java/co/cask/cdap/proto/WorkflowTokenNodeDetail.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+/**
+ * Class to represent {@link WorkflowToken} for a given workflow run at a given node.
+ */
+public class WorkflowTokenNodeDetail {
+
+  private final Map<String, String> tokenDataAtNode;
+
+  public WorkflowTokenNodeDetail(Map<String, String> tokenDataAtNode) {
+    this.tokenDataAtNode = tokenDataAtNode;
```
Comment (sagarkapare) [inline cdap-proto/src/main/java/co/cask/cdap/proto/WorkflowTokenNodeDetail.java]: should we make defensive copy here?

File: cdap-proto/src/main/java/co/cask/cdap/proto/WorkflowTokenNodeDetail.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  public Map<String, String> getTokenDataAtNode() {
+    return tokenDataAtNode;
```
Comment (sagarkapare) [inline cdap-proto/src/main/java/co/cask/cdap/proto/WorkflowTokenNodeDetail.java]: return immutable map?

**Comment by bdmogal** [comment]
@sagarkapare addressed comments and rebased with develop. please review again

File: cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java [hunk 6]
```diff
@@ -457,4 +459,22 @@ void addApplication(Id.Application id,
...
+   * Updates the {@link WorkflowToken} for a specified run of a workflow.
+   *
+   * @param workflowId {@link Id.Workflow} of the workflow whose {@link WorkflowToken} is to be updated
+   * @param workflowRunId Run Id of the workflow for which the {@link WorkflowToken} is to be updated
+   * @param token the {@link WorkflowToken} to update
+   */
+  void updateWorkflowToken(Id.Workflow workflowId, String workflowRunId, WorkflowToken token);
+
+  /**
+   * Retrieves the {@link WorkflowToken} for a specified run of a workflow
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/app/store/Store.java]: `.` at the end of the line.

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/WorkflowHttpHandler.java [hunk 7]
```diff
@@ -226,4 +245,74 @@ public void getWorkflowSchedules(HttpRequest request, HttpResponder responder,
...
+                               @PathParam("run-id") String runId,
+                               @QueryParam("scope") @DefaultValue("user") String scope,
+                               @QueryParam("key") @DefaultValue("") String key) throws NotFoundException {
+    WorkflowToken workflowToken = getWorkflowToken(namespaceId, appId, workflowId, runId);
+    WorkflowToken.Scope tokenScope = WorkflowToken.Scope.valueOf(scope.toUpperCase());
+    WorkflowTokenDetail workflowTokenDetail = WorkflowTokenDetail.of(workflowToken.getAll(tokenScope));
+    Type workflowTokenDetailType = new TypeToken<WorkflowTokenDetail>() { }.getType();
+    if (key.isEmpty()) {
+      responder.sendJson(HttpResponseStatus.OK, workflowTokenDetail,
+                         workflowTokenDetailType, GSON);
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/WorkflowHttpHandler.java]: maybe move this in single line.

File: cdap-app-fabric/src/test/java/co/cask/cdap/ConditionalWorkflowApp.java [hunk 8]
```diff
@@ -126,8 +126,12 @@ public void beforeSubmit(MapReduceContext context) throws Exception {
       FileInputFormat.addInputPath(job, new Path(inputPath));
       FileOutputFormat.setOutputPath(job, new Path(outputPath));
       // Put something in the token
-      context.getWorkflowToken().put("action_type", "MapReduce");
-      context.getWorkflowToken().put("start.time", Value.of(System.currentTimeMillis()));
+      WorkflowToken workflowToken = context.getWorkflowToken();
+      if (workflowToken == null) {
+        return;
+      }
+      workflowToken.put("action_type", "MapReduce");
```
Comment (sagarkapare) [inline cdap-app-fabric/src/test/java/co/cask/cdap/ConditionalWorkflowApp.java]: Please change this to `action.type`. I forgot to do it in my PR.

File: cdap-proto/src/main/java/co/cask/cdap/proto/WorkflowTokenDetail.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+      for (NodeValue nodeValueEntry : entry.getValue()) {
+        convertedList.add(new NodeValueDetail(nodeValueEntry));
+      }
+      converted.put(entry.getKey(), convertedList);
+    }
+    return new WorkflowTokenDetail(converted);
+  }
+
+  /**
+   * Class to represent node -> value mapping for a given key in a {@link WorkflowToken}
```
Comment (sagarkapare) [inline cdap-proto/src/main/java/co/cask/cdap/proto/WorkflowTokenDetail.java]: `.` at the end.

**Comment by sagarkapare** [comment]
Minor comments. otherwise LGTM :+1:

**Comment by bdmogal** [comment]
Fixed all comments. Will merge after a successful build.

**Comment by bdmogal** [comment]
Build Passed http://builds.cask.co/browse/CDAP-DUT2206-7. Merging.

**Pull request merged**